### PR TITLE
[DRAFT] working implementation of phase 3

### PR DESCRIPTION
**Pull request by Jovonni**
# Indexable Anchors Phase 3 - #[Issue]

## Description
Working implementation of phase 3 of the indexable anchors work

# Gas Consumption with calldata
Due to the length of the hex encoded cid, the payload is encoded the pattern `<sig>` `<location>` `<length>` `<param>` since the bytes type is dynamically sized 

calldata example, where `0x7d3b0ca2` is the function sign
```
0x7d3b0ca2000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000260f01850112204a427b02b719082b0440459d7cfad2187ee96cf3a3c2c72deac1234724b3a3460000000000000000000000000000000000000000000000000000
```

Also, the actual function paramater is padded to the right.

Include a summary of the change and which issue it addresses in the title of the PR. 

Include relevant motivation, context, brief description and impact of the change(s). List follow-up tasks here.

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [ ] Test A (e.g. Test A - New test that ... ran in local, docker, and dev unstable.)
- [ ] Test B

## Definition of Done

Before submitting this PR, please make sure:

- [ ] The work addresses the description and outcomes in the issue
- [ ] I have added relevant tests for new or updated functionality
- [ ] My code follows conventions, is well commented, and easy to understand
- [ ] My code builds and tests pass without any errors or warnings
- [ ] I have tagged the relevant reviewers
- [ ] I have updated the READMEs of affected packages
- [ ] I have made corresponding changes to the documentation
- [ ] The changes have been communicated to interested parties

## References:

Please list relevant documentation (e.g. tech specs, articles, related work etc.) relevant to this change, and note if the documentation has been updated.

### CODE REVIEW

File: .env.prod.example [hunk 1]
```diff
@@ -7,11 +7,14 @@ IPFS_API_TIMEOUT=120000
 
 BLOCKCHAIN_CONNECTOR=ethereum
 ETH_NETWORK=ganache
-ETH_GAS_LIMIT=4712388
+xETH_GAS_LIMIT=4712388
```
Comment (ukstv) [inline .env.prod.example:10]: What does this `xETH` do?

File: config/default.json [hunk 2]
```diff
@@ -1,8 +1,10 @@
 {
   "mode": "bundled",
+  "useSmartContractAnchors": false,
   "port": 8081,
   "anchorControllerEnabled": false,
-  "cronExpression": "0/5 * * * ? *",
+  "cronExpressionx": "0/5 * * * ? *",
```
Comment (ukstv) [inline config/default.json:6]: What is `cronExpressionx` here? How does it compare to `cronExpression`?

File: contracts/src/CeramicAnchorServiceV2.sol [hunk 3]
```diff
@@ -17,15 +17,24 @@ contract CeramicAnchorServiceV2 is Ownable {
 
     //upon successful anchor
     event DidAnchor(address indexed _service, bytes _root);
+    // event DidAnchor(address indexed _service, bytes32 _root);
```
Comment (ukstv) [inline contracts/src/CeramicAnchorServiceV2.sol:20]: dev comments.

In the future, please, make sure to not leave dev artefacts like this in the PRs. It is quite annoying.

File: contracts/src/CeramicAnchorServiceV2.sol [hunk 4]
```diff
@@ -17,15 +17,24 @@ contract CeramicAnchorServiceV2 is Ownable {
...
 
     // Only an address in the allow list is allowed to anchor
+    // TODO: upon contract creation, allowList is blank, so nobody is allowed
```
Comment (ukstv) [inline contracts/src/CeramicAnchorServiceV2.sol:23]: If you have a TODO like this, please, accompany it with a corresponding Linear issue, which tracks the fix.

Like this: `TODO(NET-NNNN): upon ....`

File: contracts/src/CeramicAnchorServiceV2.sol [hunk 5]
```diff
@@ -17,15 +17,24 @@ contract CeramicAnchorServiceV2 is Ownable {
...
-        require(allowList[ msg.sender ], "Allow List: caller is not allowed");
+        require(
+            ( allowList[ msg.sender ] || msg.sender == owner() ), 
+            "Allow List: caller is not allowed");
         _;
     }
 
-    constructor(){}
+    constructor(){
+        // TODO: should we add owner address to allowList upon contract creation
```
Comment (ukstv) [inline contracts/src/CeramicAnchorServiceV2.sol:36]: Again, TODO should have a corresponding Linear issue.

File: contracts/src/CeramicAnchorServiceV2.sol [hunk 6]
```diff
@@ -67,6 +75,7 @@ contract CeramicAnchorServiceV2 is Ownable {
         @desc Here _root is a byte representation of Merkle root CID.
     */
     function anchor(bytes calldata _root) public onlyAllowed {
+    // function anchor(bytes32 calldata _root1, bytes32 calldata _root2, bytes8 calldata _root3) public onlyAllowed {
```
Comment (ukstv) [inline contracts/src/CeramicAnchorServiceV2.sol:78]: It is a dev artefact, isn't it?

File: logfile [hunk 7]
```diff
@@ -0,0 +1,133 @@
+2022-06-26 10:13:19.614 EDT [82777] LOG:  starting PostgreSQL 14.2 on aarch64-apple-darwin21.3.0, compiled by Apple clang version 13.0.0 (clang-1300.0.29.30), 64-bit
```
Comment (ukstv) [inline logfile:1]: Should it be part of the PR? Should it be git-ignored?

File: package.json [hunk 8]
```diff
@@ -104,6 +105,7 @@
     "eslint": "^8.7.0",
     "eslint-config-3box": "^0.4.0",
     "eslint-plugin-jest": "^25.7.0",
+    "ganache": "^7.3.2",
```
Comment (ukstv) [inline package.json:108]: Is `ganache` a newer version of `ganache-core`? If so, maybe it makes sense to drop `ganache-core`.

File: src/__tests__/ceramic_integration.test.ts [hunk 9]
```diff
@@ -31,6 +30,12 @@ import { RequestRepository } from '../repositories/request-repository.js'
 import { Request } from '../models/request.js'
 import { CID } from 'multiformats/cid'
 
+import { ethers } from 'ethers'
+import * as fs from 'fs';
+import Ganache from 'ganache-core'
+import ganache from "ganache"
```
Comment (ukstv) [inline src/__tests__/ceramic_integration.test.ts:36]: Is it even used here?

**Review by ukstv — CHANGES_REQUESTED** [review]
Please, re-request a review, when the PR is ready.

FYI: While it is in draft, you could take someone's attention by pinging that person in a single comment on a line of interest.

File: src/__tests__/ceramic_integration.test.ts [hunk 10]
```diff
@@ -131,6 +155,9 @@ async function makeCAS(
   configCopy.ceramic.apiUrl = 'http://localhost:' + minConfig.ceramicPort
   configCopy.blockchain.connectors.ethereum.network = 'ganache'
   configCopy.blockchain.connectors.ethereum.rpc.port = minConfig.ganachePort + ''
+  //todo
```
Comment (stephhuynh18) [inline src/__tests__/ceramic_integration.test.ts:158]: You will also need to set `useSmartContractAnchors` to true

**Comment by stbrody** [comment]
@stephhuynh18 should this PR be closed?  Is there anything in here that wasn't already merged in a different PR?

**Comment by stephhuynh18** [comment]
@stbrody closing

**Pull request closed without merging**